README.md
 miniMardi 1.3 - Messy Chess Program
 -----------------------------------

This is a "oh no not another simple chess program" chess program.
It's written in C and works under Unix and Windows.

Features:
 - Opening book (small, limited by the simple search).
 - AlphaBeta search. 
 - Time management. 
 - Search heuristics (history moves).
 - Draw aware (3-fold repetition).
 - FEN support.
 - xboard compatible.
 - Zippy compatible.

Implementation Details:
 - 0x88 board representation.
 - zobrist keys for repetition detection.
 - moves encoded into an INT.
 - move sorting (mvv/lva), history moves.
 - AlphaBeta search with Quiescent.

New for this release:
 - Null move.
 - Transposition table.
 - Option for search: MTD, PV.
 - Almost no global variables.

Thanks to:
 - Bruce Moreland for his excellent site on computer chess.
 - Marcel van Kervinck, MSCP.
 - Computer Chess Club.
 - Opening book, Unknown author.

Copying:
 See file COPYING. 

   Juan Pablo Fernandez 


icsDroneICC/ChangeLog
Version 1.5.0; Wed Oct 10 16:25 METDST 2001;

* IMPORTANT:  Project renamed from RoboFICS to icsDrone and placed 
  under a BSD license.  I intend to support more servers and provide a 
  way for others to contribute more directly, so the project has found
  a home on SourceForge.net.
* Fixed a small bug in board setup as crafty doesn't support 'bd' (Thanks
  to Gian-Carlo Pascutto for reporting this)
* Detection of nickname; fixes case problems and needed for guest login
* Changed softquit/restart cmds to use gameID to determine if we are
  playing instead of seeing if the engine is running (which it would 
  always be if we're running with the '-haveCmdNew true' option)
* Reset the 'flag alarm' when a game has ended.
* Added some cmds from the Zippy v2 protocol.
* Added a special command for the FICS server to compensate for its
  poorly written move-parser.  You would think a correct move-parser
  is a fundamental part of a chess server, wouldn't you?
* Fixed bug where a 'moretime' from an opponent would send the previous 
  move to the engine one more time causing it to be confused and stall.
* It will now attempt to 'courtesyadjourn' a game where the opponent is
  3 minutes down on time and can't be flagged because of timeseal (he's
  lagging).  This will stop others having to wait forever for the opponent
  to timeout or get back.
* Fixed a potential problem with signals.  They should be safe now.
* Fixed replies to opponent when using say ([] needed to be stripped).
* Totally rewrote limitRematches code.  It now uses alarm() and gets the
  pending list from the server instead of maintaining it itself.  Currently
  it waits a minute before accepting a challenge from the same player when
  the limit has been met.
* Changed the -sendGameEnd option to send the argument following the option
  as a command to the server (instead of just 'gameend').  It also allows
  multiple commands if you seperate them by "\n" (ie. "seek 1 0\nseek 2 0")
* Changed the way it detects start of a new game.  Was nessesary as games
  started using seeks weren't correctly detected.
* Attempt to fix bad options given on commandline by interpreting some of
  the feature cmds from the engine (if it supports them).
* Now sends "force" cmd in addition to the "new" cmd to the engine when
  a game has ended.  An attempt to solve a problem with crafty hanging 
  when opponent resigned when crafty was thinking.
* New option -autoConfigure.  If set to true, it will attempt to configure
  itself by looking at the filename of the engine and set all engine-
  specific options correctly.
* New option -haveCmdPing.  If set to true, it will send a ping to the 
  engine on startup and wait for it to respond with pong before continuing.
  This allows the engine to complete slow initialization tasks (like TBs)
  without icsDrone starting a game before it's actually ready for operation.
* Added -logFile.  If logFile ends with .gz, it will use zlib to write 
  compressed data to the log.

Version 1.4.6; Thu May 21 19:45 METDST 1999;

* Security fix (thanks to Marcel van Kervinck for reporting this).
* Added "auto-flag" (thanks to Klaus Knopper for the patch).
* Fix for a theoretical buffer overflow (thanks to David Flynn for reporting
  this).
* Added support for sending various commands to the server after being
  idle for certain amount of time (check the docs).
* Added more support for remote commands; now you are able to see the 
  contents of its lists (check the docs).
* Added support for sending the result of games to the engine (check the 
  docs)
* Added support for programs that can exit on their own (particular useful
  for those that use the 'result' command to initiate learning and don't
  like to be killed by RoboFICS prematurely).
	
Version 1.4.5; Thu Sep 17 22:15 METDST 1998;

* Added fix for Crafty 15.19 and newer.  Hyatt started to use a new
  format for displaying moves that would break older winboards and
  naturally also robofics, eventhough it had no problem in the first
  place.  Anyway, with some luck maybe I'll get notified next time
  in *advance* when something like this is planned...  you think?
  If you are reading this, then please PLEASE keep this in mind - I'd
  appreciate it - thanks! :-)

Version 1.4.4; Sun Jul  5 19:00 METDST 1998;

* Owner can now give commands to the engine by prefixing it with an '*'
* Added support for 'new' command.  Tells the computer to reset all
  all data and be ready for a new game.  With this enabled, RoboFICS
  won't kill the program after each game ends - but simply send 'new'.
* Added support for 'type' command.  Tells the computer what type of
  game this is (standard, blitz, suicide, atomic, lightning, ...). 
  Useful for those computers that can play severel kinds of chess.
* Fixed it so the program argument can be a string.
* Fixed bug in move conversion.
* Fixed a bug with owner tells introduced in last bugfix :)

Version 1.4.3; Tue Apr 14 14:59 METDST 1998;

* Security fix - (thanks to Andrew Tridgell for reporting this)
* It now responds to 'says' aswell as 'tells'.
* Playing another computer could result in a loop when exchanging 'thanks'
  messages since they would keep responding to each other (thanks to 
  Marcel van Kervinck for reporting this)
* OS/2 + EMX specific StartComputer() added (thanks to William J. Little
  for the code and reading the docs :-)

Version 1.4.2; Sun Nov 30 23:43 MET 1997;

* Crafty 14.1 set a new ``standard'' for sending tells/kibitzes/etc to the
  chess server without telling anyone else on beforehand.  I now check for
  this aswell as the old "real standard" way of doing thinks to keep
  compatability with all the other chess computers.

Version 1.4.1; Tue Aug 12 22:47 METDST 1997;

* Fixed incorrect command sent to server.  It sent 'moves -l' instead of
  'moves l'.
* Fixed move convertion with regards to the 'longAlgMoves' feature.
 	
Version 1.4;  Sat Jun 21 15:25 METDST 1997;


src/board.c
#include "board.h"
/*-------------------------------------------------------.
 | file: board.c                                         |
 | contains: functions for manipulating the board.       |
 | note: the array is actually 128 elements big,         |
 |       but only 64 are used for the pieces.            |
  -------------------------------------------------------*/

/*-------------------------------------------------------.
 | function: newboard() -> board                         |
 | returns: a new 8x8 chess board                        |
  -------------------------------------------------------*/
board newboard(){
  board b;
  b = (board) malloc(sizeof(board_));
  return b;
}

/*-------------------------------------------------------.
 | function: free_board(board b) -> void                 |
 | effects: free the board b                             |
  -------------------------------------------------------*/
void free_board(board b){
  free(b);
}

/*-------------------------------------------------------.
 | function: empty_board(board b) -> void                |
 | returns: void                                         |
 | effects: sets all squares to empty on board b         |
 | note:  should not be exported.                        |
  -------------------------------------------------------*/
void empty_board(board b){
  int i=0;
  while (i<128){
    b->board[i]=EMPTY;
    i++;
  }
}

/*-------------------------------------------------------.
 | function: chess_board(board b) -> void                |
 | returns:                                              |
 | effects:  sets the board to the starting position.    |
  -------------------------------------------------------*/
void chess_board(board b){
  empty_board(b);
  /* White */
  b->board[0]=WR;
  b->board[1]=WN;
  b->board[2]=WB;
  b->board[3]=WQ;
  b->board[4]=WK;
  b->board[5]=WB;
  b->board[6]=WN;
  b->board[7]=WR;
  b->board[16]=WP;
  b->board[17]=WP;
  b->board[18]=WP;
  b->board[19]=WP;
  b->board[20]=WP;
  b->board[21]=WP;
  b->board[22]=WP;
  b->board[23]=WP;
  /* Black */
  b->board[112]=BR;
  b->board[113]=BN;
  b->board[114]=BB;
  b->board[115]=BQ;
  b->board[116]=BK;
  b->board[117]=BB;
  b->board[118]=BN;
  b->board[119]=BR;
  b->board[96]=BP;
  b->board[97]=BP;
  b->board[98]=BP;
  b->board[99]=BP;
  b->board[100]=BP;
  b->board[101]=BP;
  b->board[102]=BP;
  b->board[103]=BP;
}

/*-------------------------------------------------------.
 | function: print_board(board b) -> void                |
 | effects: prints the board b to stdout.                |
  -------------------------------------------------------*/
void print_board(board b){
  int i=112;
  piece p;
  while (i > -1) {
    if (!(i & 0x88)){
    p=b->board[i];
    switch(p) {
    case WK: {printf("wK"); break;}
    case WQ: {printf("wQ"); break;}
    case WR: {printf("wR"); break;}
    case WB: {printf("wB"); break;}
    case WN: {printf("wN"); break;}
    case WP: {printf("wP"); break;}
    case BK: {printf("bK"); break;}
    case BQ: {printf("bQ"); break;}
    case BR: {printf("bR"); break;}
    case BB: {printf("bB"); break;}
    case BN: {printf("bN"); break;}
    case BP: {printf("bP"); break;}
    case EMPTY: {printf(".."); break;}
    default: {printf("??:%i",p); break;}
    }
    i++;
    } else {
      printf("\n");
      i=i-24;
    }
  }
}
